mfa: per-session MFA certs for SSH and Kubernetes

awly opened the pull request:
This is client-side support for requesting single-use certs with an MFA
check.

The client doesn't know whether they need MFA check when accessing a
resource, this is decided during an RBAC check on the server. So a
client will always try to get a single-use cert, and the server will
respond with NotNeeded if MFA is not required. This is an extra
round-trip for every session which causes ~20% slowdown in SSH logins:

```
$ hyperfine '/tmp/tsh-old ssh talos date' '/tmp/tsh-new ssh talos date'
Benchmark #1: /tmp/tsh-old ssh talos date
  Time (mean ± σ):      49.9 ms ±   1.0 ms    [User: 15.1 ms, System: 7.4 ms]
  Range (min … max):    48.4 ms …  54.1 ms    59 runs

Benchmark #2: /tmp/tsh-new ssh talos date
  Time (mean ± σ):      60.2 ms ±   1.6 ms    [User: 19.1 ms, System: 8.3 ms]
  Range (min … max):    59.0 ms …  69.7 ms    50 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Summary
  '/tmp/tsh-old ssh talos date' ran
    1.21 ± 0.04 times faster than '/tmp/tsh-new ssh talos date'
```

Another few other internal changes:

- client.LocalKeyAgent will now always have a non-nil LocalKeyStore.
  Previously, it would be nil (e.g. in a web UI handler or when using an
  identity file) which easily causes panics. I added a noLocalKeyStore
  type instead that returns errors from all methods.

- requesting a user cert with a TTL < 1min will now succeed and return a
  1min cert instead of failing

Commit:
mfa: per-session MFA certs for SSH and Kubernetes

This is client-side support for requesting single-use certs with an MFA
check.

The client doesn't know whether they need MFA check when accessing a
resource, this is decided during an RBAC check on the server. So a
client will always try to get a single-use cert, and the server will
respond with NotNeeded if MFA is not required. This is an extra
round-trip for every session which causes ~20% slowdown in SSH logins:

```
$ hyperfine '/tmp/tsh-old ssh talos date' '/tmp/tsh-new ssh talos date'
Benchmark #1: /tmp/tsh-old ssh talos date
  Time (mean ± σ):      49.9 ms ±   1.0 ms    [User: 15.1 ms, System: 7.4 ms]
  Range (min … max):    48.4 ms …  54.1 ms    59 runs

Benchmark #2: /tmp/tsh-new ssh talos date
  Time (mean ± σ):      60.2 ms ±   1.6 ms    [User: 19.1 ms, System: 8.3 ms]
  Range (min … max):    59.0 ms …  69.7 ms    50 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Summary
  '/tmp/tsh-old ssh talos date' ran
    1.21 ± 0.04 times faster than '/tmp/tsh-new ssh talos date'
```

Another few other internal changes:

- client.LocalKeyAgent will now always have a non-nil LocalKeyStore.
  Previously, it would be nil (e.g. in a web UI handler or when using an
  identity file) which easily causes panics. I added a noLocalKeyStore
  type instead that returns errors from all methods.

- requesting a user cert with a TTL < 1min will now succeed and return a
  1min cert instead of failing

a-palchikov reviewed (approved):
I'm not certain about other ramifications of MFA check but if the 20% slowdown is a concern, the connect/port forward APIs could attempt to connect with the local agent's ssh key first (w/o the embedded node name) before requesting the certificate with MFA check if RBAC requires it.

russjones commented:
@fspmarshall Can you review this?

fspmarshall commented on lib/client/client.go:
```diff
@@ -1117,3 +1260,24 @@ func (proxy *ProxyClient) currentCluster() (*services.Site, error) {
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	key, err := proxy.IssueUserCertsWithMFA(ctx, ReissueParams{
+		NodeName:       node,
+		RouteToCluster: cluster.Name,
+	})
```
In order for this to work with the access request system, the request IDs need to be propagated (`ReissueParams.AccessRequests` field).  IDs of the currently active requests are stored on the SSH certificate.  #5441 introduces a helper method for extracting them, but you can do it manually with something like this:
```golang
var activeRequests services.RequestIDs
rawRequests, ok := sshCert.Extensions[teleport.CertExtensionTeleportActiveRequests]
if ok {
	if err := activeRequests.Unmarshal([]byte(rawRequests)); err != nil {
		return activeRequests, trace.Wrap(err)
	}
}
key, err := proxy.IssueUserCertsWithMFA(ctx, ReissueParams{
	AccessRequests: activeRequests.AccessRequests,
	// ...
})
```

awly replied:
Thanks for the catch, added inside `IssueUserCertsWithMFA` to cover more callers.

awly commented on lib/auth/grpcserver.go:
```diff
@@ -1603,26 +1618,115 @@ func (g *GRPCServer) GenerateUserSingleUseCerts(stream proto.AuthService_Generat
+			if err != nil {
+				addr = n.GetAddr()
+			}
+			// Match NodeName to UUID, hostname or self-reported server address.
+			if n.GetName() == req.NodeName || n.GetHostname() == req.NodeName || addr == req.NodeName {
+				node = n
+				break
+			}
```
`n.GetName()` returns the UUID, so the first condition in this `if` statement matches nodes by UUID

fspmarshall replied:
Yeah, what I mean is that iteration shouldn't halt if a hostname/addr match is found because there might still be a node which matches UUID and that would need to take priority.

awly replied:
If some node has a hostname matching other node's UUID (which seems very unlikely), then we have multiple matches and no way to resolve the conflict and for the user to pick one of them.

I think the more likely issue is when user passes a hostname and multiple nodes match.
Changed the logic to catch that scenario - if there are multiple matches return an error.

Commit:
Capture access approvals on MFA-issued certs

awly commented:
PTAL @fspmarshall

Commit:
Address review feedback

awly commented on lib/auth/grpcserver.go:
```diff
@@ -1612,26 +1627,117 @@ func (g *GRPCServer) GenerateUserSingleUseCerts(stream proto.AuthService_Generat
+	}
+	// Errors other than ErrSessionMFARequired mean something else is wrong,
+	// most likely access denied.
+	if noMFAAccessErr != services.ErrSessionMFARequired {
```
good point.
I used `trace.Unwrap` originally and had trouble (because I think `trace.AccessDenied` unwraps into the underlying string error). But `errors.Is` works well, updated.

awly commented:
@r0mant @russjones please review

r0mant commented on lib/auth/grpcserver.go:
```diff
@@ -1612,26 +1627,116 @@ func (g *GRPCServer) GenerateUserSingleUseCerts(stream proto.AuthService_Generat
+// authentication is not required for this session. Any other error means
+// failed validation.
+func validateUserSingleUseCertRequest(ctx context.Context, actx *grpcContext, req *proto.UserCertsRequest) error {
+	if req.Username != actx.User.GetName() {
```
I wonder if you can use `actx.currentUserAction(req.Username)` here to not spread this authz logic across different servers (grpc vs server with roles)?

r0mant commented on lib/auth/grpcserver.go:
```diff
@@ -1612,26 +1627,116 @@ func (g *GRPCServer) GenerateUserSingleUseCerts(stream proto.AuthService_Generat
+			}
+		}
+		if len(matches) == 0 {
+			return trace.NotFound("target resource not found")
```
Is the resource type/name not revealed in these messages on purpose?

awly replied:
I was just being lazy, updated

r0mant commented on lib/auth/grpcserver.go:
```diff
@@ -1612,26 +1627,116 @@ func (g *GRPCServer) GenerateUserSingleUseCerts(stream proto.AuthService_Generat
+// failed validation.
+func validateUserSingleUseCertRequest(ctx context.Context, actx *grpcContext, req *proto.UserCertsRequest) error {
+	if req.Username != actx.User.GetName() {
+		return trace.BadParameter("cannot request a single-use certificate for a different user")
```
trace.AccessDenied?

r0mant commented on lib/client/api.go:
```diff
@@ -1024,6 +1024,16 @@ func (tc *TeleportClient) ReissueUserCerts(ctx context.Context, params ReissuePa
 	return proxyClient.ReissueUserCerts(ctx, params)
 }
 
+func (tc *TeleportClient) IssueUserCertsWithMFA(ctx context.Context, params ReissueParams) (*Key, error) {
```
Nit: Add godoc.

r0mant commented on lib/client/mfa.go:
```diff
@@ -42,7 +42,7 @@ func PromptMFAChallenge(ctx context.Context, proxyAddr string, c *proto.MFAAuthe
 	// TOTP only.
 	case c.TOTP != nil && len(c.U2F) == 0:
-		totpCode, err := prompt.Input(os.Stdout, os.Stdin, fmt.Sprintf("Enter an OTP code from a %sdevice", promptDevicePrefix))
+		totpCode, err := prompt.Input(os.Stderr, os.Stdin, fmt.Sprintf("Enter an OTP code from a %sdevice", promptDevicePrefix))
```
Just curious why this change to stderr?

awly replied:
Because of `kubectl`'s exec plugins.
We configure `kubectl` to exec a `tsh` subcommand every time it needs to get credentials to connect.
`kubectl` expects to read the credentials from `stdout` of the execed process.
Anything written to `stderr` will be forwarded to `kubectl`'s own `stderr` to allow interactive commands like this.

r0mant commented on lib/client/mfa.go:
```diff
@@ -126,6 +126,7 @@ func promptU2FChallenges(ctx context.Context, proxyAddr string, challenges []*pr
 		})
 	}
 
+	log.Debugf("prompting U2F devices with facet %q", facet)
```
Feels like a debug leftover, need this?

awly replied:
This is left on purpose.
If a user mis-configures U2F facets (proxy addresses, basically) on the `auth_service`, it's really hard to debug when they use a different value in `tsh --proxy=...` and U2F auth will fail.

Commit:
Address review feedback

Commit:
mfa: accept unknown nodes during short-term MFA cert creation

An unknown node could be an OpenSSH node set up via
https://goteleport.com/teleport/docs/openssh-teleport/

In this case, we shouldn't prevent the user from connecting.

There's a small risk of authz bypass - an attacker might know a
different name/IP for a registered node which Teleport doesn't know
about. But a Teleport node will still check RBAC and reject the
connection.

Commit:
Validate username against unmapped user identity

IssueUserCertsWithMFA is called on the leaf auth server in case of
trusted clusters. Username in the request object will be that of the
original unmapped caller.

Commit:
mfa: add IsMFARequired RPC

This RPC is ran before every connection to check whether MFA is
required. If a connection is against the leaf cluster, this request is
forwarded from root to leaf for evaluation.

awly commented:
Added `IsMFARequired` RPC on the auth server.
The client now uses this against the appropriate auth server (root or leaf) to check if MFA is required.
If it is, client sends the `UserSingleUseCerts` RPC to the **root** cluster, which should always be the one issuing credentials.

@r0mant @fspmarshall PTAL

fspmarshall commented on api/client/proto/authservice.proto:
```diff
@@ -709,6 +709,25 @@ message UserSingleUseCertsResponse {
+// IsMFARequiredRequest is a request to check whether MFA is required to access
+// the Target.
+message IsMFARequiredRequest {
+    // Username of the caller.
+    string Username = 1;
```
Clarify that this is OS user for target host, not teleport username.

Commit:
Fix integration tests

r0mant commented on lib/auth/auth_with_roles.go:
```diff
@@ -66,7 +68,10 @@ func (a *ServerWithRoles) action(namespace string, resource string, action strin
+	if hasLocalUserRole(a.context.Checker) && username == a.context.Identity.GetIdentity().Username {
+		return nil
+	}
+	if hasRemoteUserRole(a.context.Checker) && username == a.context.UnmappedIdentity.GetIdentity().Username {
```
What's the implication of this change? Wouldn't this mean that, for example, a remote user has permissions to change the local user's password as long as they have the same username? I may be missing something.

awly replied:
Nice catch, didn't think about that.
It also turns out I was confusing "username" with "ssh login name", so I renamed that request field and it's no longer compared to the caller username.
Changes to `currentUserAction` reverted.

r0mant commented on lib/auth/auth_with_roles.go:
```diff
@@ -2601,6 +2611,132 @@ func (a *ServerWithRoles) GenerateUserSingleUseCerts(ctx context.Context) (proto
+	}
+
+	var noMFAAccessErr, notFoundErr error
+	switch t := req.Target.(type) {
```
More of a meta comment, but don't we want to keep the ACL layer slim without much business logic? Any downsides of moving this switch downstream, e.g. to an auth server method and invoke it here similar to how other methods here do? You can pass checker to it too.

awly replied:
All of this code felt like pure authz logic, which is why I put it here.
But I think you're right, moved all of this code into `auth.Server`.

Commit:
Correctly treat "Username" as login name in IsMFARequired

Also, move the logic into auth.Server out of ServerWithRoles.

Commit:
Fix TestHA

Commit:
Merge branch 'master' into andrew/session-2fa-cli

r0mant commented on lib/client/client.go:
```diff
@@ -203,10 +246,156 @@ func (proxy *ProxyClient) ReissueUserCerts(ctx context.Context, params ReissuePa
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	tlsCert, err := key.TeleportTLSCertificate()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	rootClusterName, err := tlsca.ClusterName(tlsCert.Issuer)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
```
Nit: There's `RootClusterName()` method on the ProxyClient which you could probably just use instead of extracting from the TLS cert, although the outcome would probably be the same.

r0mant commented on lib/client/client.go:
```diff
@@ -203,10 +246,156 @@ func (proxy *ProxyClient) ReissueUserCerts(ctx context.Context, params ReissuePa
+	}
+
+	// Note: always connect to root for getting new credentials.
+	clt, err := proxy.ConnectToCluster(ctx, params.RouteToCluster, true)
```
Shouldn't we use `proxy.ConnectToRootCluster()` here if we only ever want to connect to root? Or is the comment inaccurate?

r0mant commented on lib/client/client.go:
```diff
@@ -203,10 +246,156 @@ func (proxy *ProxyClient) ReissueUserCerts(ctx context.Context, params ReissuePa
+
+	// Always connect to root for getting new credentials, but attempt to reuse
+	// the existing client if possible.
+	if params.RouteToCluster != rootClusterName {
```
If the comment above about connecting to root is accurate, then this is probably not needed?

r0mant commented on lib/client/client.go:
```diff
@@ -203,10 +246,156 @@ func (proxy *ProxyClient) ReissueUserCerts(ctx context.Context, params ReissuePa
+		return nil, trace.Wrap(err)
+	}
+	certResp := resp.GetCert()
+	if cert == nil {
```
Should this be `if certResp == nil`?

r0mant reviewed (approved):
Couple more questions but otherwise lgtm.

Commit:
Address review feedback

Commit:
Merge branch 'master' into andrew/session-2fa-cli
